pandas/core/case_when.py

```diff
@@ -0,0 +1,210 @@
+        `replacement` should be a 1-D array or a scalar.
+        If `replacement` is a 1-D array, it should have the same
+        shape as the paired `condition`.
+        When multiple boolean conditions are satisfied,
+        the first replacement is used.
+
+    default : scalar, array-like, default None
+        If provided, it is the replacement value to use
+        if all conditions evaluate to False.
+        If default is a 1-D array, it should have the same shape as
+        `condition` and `replacement`. If not specified,
+        entries will be filled with the corresponding
+        NULL value (``np.nan`` for numpy dtypes, ``pd.NA`` for extension
+        dtypes).
+
+    level : int, default None
+        Alignment level if needed.
+
+        .. versionadded:: 2.2.0
+
+    Returns
+    -------
+    Series
+
+    See Also
+    --------
+    Series.mask : Replace values where the condition is True.
+
+    Examples
+    --------
+    >>> df = pd.DataFrame({
+    ...     "a": [0,0,1,2],
+    ...     "b": [0,3,4,5],
+    ...     "c": [6,7,8,9]
+    ... })
+    >>> df
+       a  b  c
+    0  0  0  6
+    1  0  3  7
+    2  1  4  8
+    3  2  5  9
+
+    >>> pd.case_when(df.a.gt(0), df.a,   # condition, replacement
+    ...              df.b.gt(0), df.b,
+    ...              default=df.c)        # optional
+    0    6
+    1    3
+    2    1
+    3    2
+    Name: c, dtype: int64
+    """
+    from pandas import Series
+
+    # summary of logic -
+    # get booleans and replacements
+    # and iteratively get the final output via Series.mask
+    # iteration starts from the end,
+    # ensuring that if there are any booleans with the same replacement,
+    # the first one is taken
+    # as much as possible the computation is pushed to Series.mask
+    len_args = len(args)
+    if not len_args:
+        raise ValueError(
+            "Kindly provide at least one boolean condition, "
+            "with a corresponding replacement."
+        )
+    if len_args % 2:
+        raise ValueError(
+            "The number of boolean conditions should be equal "
+            "to the number of replacements. "
+            "However, the total number of conditions and replacements "
+            f"is {len(args)}, which is an odd number."
+        )
+
+    counter = len_args // 2 - 1
+    counter = range(counter, -1, -1)
+    conditions = []
+    for num, condition in zip(counter, args[-2::-2]):
+        if not hasattr(condition, "shape"):
+            condition = np.asanyarray(condition)
+        if condition.ndim > 1:
+            raise ValueError(f"condition{num} is not a one dimensional array.")
+        if not is_bool_dtype(condition):
+            raise TypeError(f"condition{num} is not a boolean array.")
+        conditions.append(condition)
+    bool_length = {condition.size for condition in conditions}
+    if len(bool_length) > 1:
+        raise ValueError("All boolean conditions should have the same length.")
```

Comment:
I think we should respect alignment

Reply:
@rhshadrach could you elaborate pls?

Reply:
Most operations align on indices.

```
value = pd.DataFrame({'a': [1, 1, 2], 'b': [3, 4, 5], 'c': [6, 7, 8]}).set_index(['a', 'b'])['c']
mask = pd.Series([True, False], index=pd.Index([1, 2], name='a'))
print(value)
# a  b
# 1  3    6
#    4    7
# 2  5    8
# Name: c, dtype: int64

print(value.mask(mask, 0))
# a  b
# 1  3    0
#    4    0
# 2  5    8
# Name: c, dtype: int64
```

Reply:
Ahhh ... I assumed wrongly that booleans were excluded from alignment. Gotcha

Reply:
Also - I believe you can remove these checks and everything else will just work. If you run into difficulties, let me know and I can also have a look.